cpp/src/arrow/type.h

```diff
@@ -116,7 +116,7 @@ struct DataType {
 
   bool Equals(const DataType* other) {
     // Call with a pointer so more friendly to subclasses
-    return this == other || (this->type == other->type);
+    return this == other || ((other != nullptr) && (this->type == other->type));
```

Comment:
agree.  And so does the compiler actually, when I tried adding in the this != nullptr, it complains that this should never be null in well defined c++ code.  I switched it to your suggestion.